### PR TITLE
fixing dataset downloads page, adding missing papers/graphs.py file

### PR DESCRIPTION
**Pull request by vsoch**
This pull request will address a few issues with the download cart.

 - #115: to fix this the easiest thing was to not use the server rendered table (which has add/remove as ajax) and switch to using a page refresh. We have to do this because I couldn't figure out a way to update the request.session without doing this. Another option was to implement a function that returned a new list of dataset ids and metadata to re-render the table, but that's hugely complex when a page refresh works perfectly.
 - #109: this should now work - we do this by clearing the table in the browser, and then we clear the session on the server (so the cart is indeed empty on the next refresh)
 - #95 I could not reproduce - we should figure out if it's still an issue.

Also, the papers/graphs.py file was missing, adding it here!

Signed-off-by: vsoch <vsochat@stanford.edu>

### CODE REVIEW

**Comment by vsoch** [comment]
Deployed to staging.

**Comment by abarysh** [comment]
Testing this on staging -- the "Add" dataset to the Download cart (e.g., at /papers/289/ and /datasets/?query=HU) doesn't seem to be working now. (So I can't really test the other fixes if I can't add datasets to the cart)

**Comment by vsoch** [comment]
That's weird, it still works for me:

image

I didn't touch any files relevant to either of those views, so I'm not sure what might be up! Is there something else I need to do to reproduce the issue you are having? Is there any error in the console?

**Comment by vsoch** [comment]
Is your cart filled to the max? That's something else we needed to fix (not in this PR) is delivering that message to the user - if you go above the limit (which I think we set to 50) you wouldn't be able to add any more.

**Comment by abarysh** [comment]
Try this:
- Clear your Download Cart
- Go to Papers and search for Parsons
- Go to Parsons~Boone, 2006 (should have ~80 datasets)
- Start adding the datasets
- At some point during the process, add and remove the same dataset (or different datasets) multiple times
- Switch from across pages a few times
- At some point during this process, the Add buttons start glitching (need to click twice to add or remove)
- If I go to Download Cart, clear the Cart and come back to this paper, I can no longer add anything (if I click on Add, nothing happens).

Interesting note: Once I "break" the add buttons, I can't reset the website to start working again. Even if I close the browser, the Add buttons still don't work. I had to switch to Chrome to Safari to try to reproduce the bug, because Chrome wouldn't let me add anything anymore. Clearing the cookies didn't seem to help.

**Comment by vsoch** [comment]
Could it be that something in the response is just breaking the javascript for the page? This is what I see in my terminal:

image

I'm not doing any more work for the evening (dinner time!) but I'll look again tomorrow. What's confusing is that I didn't change anything relevant to that particular page in this PR, so I'm not sure why it's doing that. But I don't see the error in production so there must be something different. Let me know if you have any insights!

**Comment by vsoch** [comment]
And geez, I just reloaded the page and it worked fine this time! So I think we have some kind of transient error. Maybe the delay is the browser trying to keep up with multiple server clicks / requests, or something like that?

**Comment by vsoch** [comment]
okay, I've been working on this all morning, and I understand the problem and I think I might have fixed it or at least helped a bit - it seems less buggy now. Also I want to note that this bug does appear to be present in production (in the same way that it appeared some times for me on staging, some times not, when I clicked enough I saw the same behavior in production) so it's unrelated to the changes in this PR, but we might as well try to fix here. The issue is likely two fold:

1. We were adding the event again each time the button is re-rendered, possibly this means clicking once triggers multiple times, and this would increase with more clicks.
2. With rapid clicking, the user is able to click the button more than once before the first request is done, this leads to the buggy flashing that I've seen when I click faster.

So what I did is to ensure that the button is disabled when the click starts, and undisabled after - this will prevent the user from clicking multiple times and triggering the event before the previous response has had a chance to finish. I also unbind the previous click events before adding a new one (basically when we remove the class that defines the button functionality and define a new class, we have to re-tag the event listener. The other option would be to render two buttons and then hide/display, but that would have it's own issues.

Finally, I moved the cart functions into a common file that can be shared by the server side and browser side rendered tables.

Deployed to staging.

File: yeastphenome/apps/common/templates/base/base.html [hunk 1]
```diff
@@ -60,6 +60,7 @@
 
     </body>
     {% endblock %}
+    {% block topscripts %}{% endblock %}
```
Comment (vsoch) [inline yeastphenome/apps/common/templates/base/base.html:63]: We can remove these - I was adding them to test something (controlling the position of where a script is inserted) but wound up not needing them and forgot to remove them.

**Comment by abarysh** [comment]
I'm not sure if this is the same problem, but it's still not completely functional for me:

- I have an empty Download Cart, I go to /papers/289/ and start adding datasets (I click relatively slowly, no more than 1 dataset per second)
- After 40, I check the download cart to make sure it looks good and yes, looks good, has 40 datasets in it
- At dataset 47, the buttons stop working -- I click on Add, it turns a lighter shade of blue (looks inactive) and doesn't turn to "Remove". 
- If I go back, remove the previous dataset and click Add again, I get the same problem again -- inactive button.

After waiting some time (~5 min), I refreshed the page and the buttons are working again.

**Comment by vsoch** [comment]
The button is made inactive while it's waiting for the server response, so if it never gets active I suspect the response is slow or not received? I also just thought of another issue - if something depends on session data (e.g., the cart) since we aren't refreshing the page and are changing the session I'm not sure what kind of bugs that might lead to.

This kind of action is really more appropriate for something reactive like react or vue.js. The solution here to have jquery binding / unbinding every button is just not very good (and this is my doing). Should we rethink how this works? Maybe remove the single add buttons and restore the button to add all of the datasets to the cart? Or disable the user being able to remove from the cart from the page (so they basically can click to add, and then it's added and the button is disabled?) We could also have a row of checkboxes, and the user could select all and then click "Add to cart" or "Remove from cart" buttons at the bottom of the page. It would require a page reload to update the cart, either way.

Feel free to take a shot at this, because I'm epic-ally failing!

**Comment by abarysh** [comment]
Ok, let's move this to a discussion issue (I'll start one now) and think about this issue more globally. In the meantime, I think it's ok to deploy or reverse these changes (whatever is easier). We'll address this issue once we figure out what we want as a whole.

**Comment by vsoch** [comment]
What do you mean reverse? The bug is also present in our current master/production, but the version here might be a slight improvement to disable the buttons and also better organize files. I'm not sure how to reverse because I'm not sure we ever had a correctly working state.

**Comment by abarysh** [comment]
Yeah, that's roughly what i meant -- since we had the problem before and we don't have a solution here, we can either go back to our current production state or (if we think the current state is slightly better) deploy the changes we made (even if they don't fully solve the problem). It sounds like going ahead with the deployment is the better option.

**Comment by vsoch** [comment]
Ah, gotcha! Yes I do think it's slightly better, so we will deploy! I'm going for a run (it's getting dark already, I've sort of waited too long today and I'll be coming back in the dark) but I'll deploy when I get back! And the more I think about it, I think having checkboxes with single "Add" and "Remove" buttons at the bottom might work better for this.

**Comment by vsoch** [comment]
Deployed to production.

**Pull request merged**